File: src/web/magmac/app/LexingStage.ts
import { Result } from "../../magmac/api/result/Result";
import { CompileResult } from "../../magmac/app/compile/error/CompileResult";
import { Node } from "../../magmac/app/compile/node/Node";
import { ApplicationError } from "../../magmac/app/error/ApplicationError";
import { ThrowableError } from "../../magmac/app/error/ThrowableError";
import { Sources } from "../../magmac/app/io/sources/Sources";
import { Lexer } from "../../magmac/app/stage/lexer/Lexer";
import { UnitSet } from "../../magmac/app/stage/unit/UnitSet";
import { Function } from "../../java/util/function/Function";
export class LexingStage<T> {
	getUnitSetApplicationErrorResult(sources1 : Sources) : Result<UnitSet<T>, ApplicationError> {return 0.readAll( ).mapErr( 0.new).mapErr( 0.new).flatMapValue( 0.getUnitSetApplicationErrorResult);;}
	getUnitSetApplicationErrorResult(units : UnitSet<String>) : Result<UnitSet<T>, ApplicationError> {return 0.lexer.apply( 0).flatMapValue( 0).toResult( ).mapErr( 0.new);;}
}

File: packages/compiler-java/src/main/java/magmac/app/LexingStage.java
package magmac.app;

import magmac.api.result.Result;
import magmac.app.compile.error.CompileResult;
import magmac.app.compile.node.Node;
import magmac.app.error.ApplicationError;
import magmac.app.error.ThrowableError;
import magmac.app.io.sources.Sources;
import magmac.app.stage.lexer.Lexer;
import magmac.app.stage.unit.UnitSet;

import java.util.function.Function;

public record LexingStage<T>(Lexer lexer, Function<Node, CompileResult<T>> deserializer) {
    public Result<UnitSet<T>, ApplicationError> getUnitSetApplicationErrorResult(Sources sources1) {
        return sources1.readAll()
                .mapErr(ThrowableError::new)
                .mapErr(ApplicationError::new)
                .flatMapValue(this::getUnitSetApplicationErrorResult);
    }

    private Result<UnitSet<T>, ApplicationError> getUnitSetApplicationErrorResult(UnitSet<String> units) {
        return this.lexer.apply(units)
                .flatMapValue((UnitSet<Node> roots) -> roots.mapValuesToResult(this.deserializer))
                .toResult()
                .mapErr(ApplicationError::new);
    }
}